packages/agreement-process/src/services/readModelService.ts

```diff
@@ -318,51 +319,108 @@ async function searchTenantsByName(
 }
 
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-function getDelegateAgreementsFilters(producerIds: TenantId[] | undefined) {
-  return producerIds && producerIds.length > 0
-    ? [
+const getDelegationLookup = (kind: DelegationKind) => [
+  {
+    $lookup: {
+      from: "delegations",
+      let: {
+        eserviceId: "$data.eserviceId",
+        consumerId: "$data.consumerId",
+        producerId: "$data.producerId",
+      },
+      pipeline: [
         {
-          $lookup: {
-            from: "delegations",
-            localField: "data.eserviceId",
-            foreignField: "data.eserviceId",
-            as: "delegations",
+          $match: {
+            $expr: {
+              $and: [
+                {
+                  $eq: ["$data.eserviceId", "$$eserviceId"],
+                },
+                {
+                  $eq: [
+                    "$data.delegatorId",
+                    kind === delegationKind.delegatedConsumer
+                      ? "$$consumerId"
+                      : "$$producerId",
```

Comment:
Let's use an exhaustive TS pattern matching, so that if for some reason we add delegation kinds in the future we get warned by TS